component/src/main/java/org/ballerinalang/sap/nativeimpl/consumer/IDocHandlerFactory.java

```diff
@@ -0,0 +1,158 @@
+ */
+public class IDocHandlerFactory implements JCoIDocHandlerFactory {
+
+    private static Log log = LogFactory.getLog(IDocHandlerFactory.class);
+    private Map<String, IDocHandlerBase> handlersMap;
+    private Map<String, Resource> sapService;
+
+    public IDocHandlerFactory(Map<String, Resource> sapService) {
+
+        handlersMap = new HashMap<String, IDocHandlerBase>();
+        this.sapService = sapService;
+        log.info("......IDocHandlerFactory is created.");
+    }
+
+    public boolean registerHadler(String idocType, IDocHandlerBase handler) {
+
+        handlersMap.put(idocType, handler);
+        log.info("For the IDoc type=" + idocType + " the handler " + handler.getClass() + " is registered");
+        return true;
+    }
+
+    @Override
+    public JCoIDocHandler getIDocHandler(JCoIDocServerContext serverCtx) {
+
+        return new JCoIDocHandlerImplementation(sapService);
+    }
+
+    class JCoIDocHandlerImplementation implements JCoIDocHandler {
+
+        private ResponseCallback callback;
+        private IDocXMLProcessor xmlProcessor;
+        private Map<String, Resource> sapService;
+
+        public JCoIDocHandlerImplementation(Map<String, Resource> sapService) {
+
+            this.sapService = sapService;
+            callback = new ResponseCallback();
+            this.xmlProcessor = JCoIDoc.getIDocFactory().getIDocXMLProcessor();
+        }
+
+        @Override
+        public void handleRequest(JCoServerContext serverCtx, IDocDocumentList idocList) {
+
+            CustomServerTIDHandler tidh = (CustomServerTIDHandler) serverCtx.getServer().getTIDHandler();
+            if (log.isDebugEnabled()) {
+                log.debug("New IDoc received");
+            }
+            try {
+                Map<String, Object> properties = new HashMap<>();
+                Resource onMessageResource = sapService.get(RESOURCE_ON_MESSAGE);
+                ProgramFile programFile = onMessageResource.getResourceInfo().getPackageInfo().getProgramFile();
+                BMap<String, BValue> consumerStruct = BLangConnectorSPIUtil.createBStruct(programFile,
+                        SAP_NATIVE_PACKAGE, CONSUMER_STRUCT_NAME);
+                log.info("idocList type: " + idocList.getIDocType());
+                Executor.submit(onMessageResource, callback, properties, null,
+                        getSignatureParameters(idocList, consumerStruct));
+            } catch (BallerinaConnectorException e) {
+                SapUtils.invokeOnError(sapService, callback, e.getMessage());
+            }
+        }
+
+        private BMap<String, BValue> getSignatureParameters(IDocDocumentList idocList,
+                                                            BMap<String, BValue> consumerStruct) {
+
+            IDocXMLProcessor xmlProcessor = JCoIDoc.getIDocFactory().getIDocXMLProcessor();
+            OutputStreamWriter osw = null;
+            ByteArrayOutputStream baos = new ByteArrayOutputStream();
+            try {
```

Comment:
Can we use `try` block with resources so that we can eliminate the finally block? 
i.e: 
```ballerina
try (OutputStreamWriter osw = new OutputStreamWriter(baos, StandardCharsets.UTF_8)) {
    // logic goes here
} catch (IOException e) { 
   // Exception handling logic
}
```